crest/Assets/Crest/Crest/Scripts/Shapes/ShapeGerstnerSplineHandling.cs

```diff
@@ -142,71 +161,86 @@ public static bool GenerateMeshFromSpline<SplinePointCustomData>(Spline.Spline s
                 }
-                else
+                for (int i = 1; i < sampledPtsOffSplineRight.Length - 1; i++)
                 {
-                    for (var j = 0; j < smoothingIterations; j++)
-                    {
-                        for (var i = 0; i < sampledPtsOffSpline.Length; i++)
-                        {
-                            // Slightly odd indexing. The first and last point are the same, the indices need to wrap to either the
-                            // second element (if overflow) or the penultimate element (if underflow) to ensure tension is maintained at ends.
-                            var ibefore = i - 1;
-                            var iafter = i + 1;
-
-                            if (ibefore < 0) ibefore = sampledPtsOffSpline.Length - 2;
-                            if (iafter >= sampledPtsOffSpline.Length) iafter = 1;
-
-                            scratchPoints[i] = (sampledPtsOffSpline[i] + sampledPtsOffSpline[iafter] + sampledPtsOffSpline[ibefore]) / 3f;
-                            scratchPoints[i] = sampledPtsOnSpline[i] + (scratchPoints[i] - sampledPtsOnSpline[i]).normalized * radius * radiusMultiplier[i];
-                        }
-                        var tmp = sampledPtsOffSpline;
-                        sampledPtsOffSpline = scratchPoints;
-                        scratchPoints = tmp;
-                    }
+                    sampledPtsOffSplineRight[i] = sampledPointsScratch[i];
                 }
             }
 
-            return UpdateMesh(transform, sampledPtsOnSpline, sampledPtsOffSpline, customData, spline._closed, ref mesh);
+            return UpdateMesh(transform, sampledPtsOffSplineLeft, sampledPtsOffSplineRight, customData, spline._closed, ref mesh);
+        }
+
+        static void ResolveOverlaps(ref Vector3[] points, Vector3[] pointsOnSpline)
+        {
+            var pointsNew = new Vector3[points.Length];
+            for (int i = 0; i < 1; i++)
+            {
+                pointsNew[i] = points[i];
+            }
+
+            Vector3 lastGoodPoint = points[Mathf.Min(points.Length - 1, 1)];
+            for (int i = 1; i < points.Length; i++)
+            {
+                var tangentSpline = pointsOnSpline[i] - pointsOnSpline[i - 1];
+                var tangent = points[i] - lastGoodPoint;
+
+                tangent.y = tangentSpline.y = 0f;
+
+                float dp = Vector3.Dot(tangent, tangentSpline);
+
+                if (dp > 0f)
+                {
+                    pointsNew[i] = points[i];
+                    lastGoodPoint = points[i];
+                }
+                else
+                {
+                    pointsNew[i].x = lastGoodPoint.x;
+                    pointsNew[i].y = points[i].y;
+                    pointsNew[i].z = lastGoodPoint.z;
+                }
+            }
+            points = pointsNew;
         }
```

Comment:
Some comments here will be helpful. Even small stuff like "Copying spline points" can make things easier.

Reply:
yeah sorry, documented and tidied a couple of bits

Reply:
Thanks.